include/revng/Support/GraphAlgorithms.h

```diff
@@ -398,15 +398,42 @@ nodesBetweenReverse(GraphT Source, GraphT Destination) {
   return nodesBetweenImpl<GraphT, Inverse<GraphT>>(Source, Destination);
 }
 
-template<class GraphT>
+/// Helper function which checks that the `ScopeGraph` is a
+template<class GraphT, class NodeT>
 bool isDAG(GraphT Graph) {
-  for (llvm::scc_iterator<GraphT> I = llvm::scc_begin(Graph),
-                                  IE = llvm::scc_end(Graph);
-       I != IE;
-       ++I) {
-    if (I.hasCycle())
-      return false;
+
+  using NodeRef = llvm::GraphTraits<GraphT>::NodeRef;
+
+  // We start by collecting all the nodes in the graph. We then remove from this
+  // set, at each iteration, each node composing the `scc` that we find. We do
+  // this until no node remains in the set, meaning that we visited all the
+  // nodes in the graph, also eventual nodes, disconnected from the entry, which
+  // may form a loop.
+  llvm::SmallSet<NodeRef, 4> RemainingNodes;
```

Comment:
The graph is not being mutated.
Just use a set of `VisiteNodes`. 
Then instead of `while (not RemainingNodes.empty())` just iterate on `nodes(Graph)` and to the SCC visit inside, always adding `I*` to `VistedNodes`.
In this way you only do a single visit of the graph instead of 2.

Reply:
See here https://github.com/revng/revng/blob/develop/lib/DataLayoutAnalysis/Middleend/CollapseSCC.cpp#L42

Reply:
Yeah, so we want to use the `VisitedNodes` as a skiplist, gotcha, integrated the changes here https://github.com/revng/revng/blob/855a9309ec0a5bda6f3236bd98f3000a6d7cb36d/include/revng/Support/GraphAlgorithms.h#L412